src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -956,6 +1076,64 @@ public DocumentId AddTextContainer(SourceTextContainer textContainer, string ful
                 return documentId;
             }
 
+            public void AddDynamicFile(IDynamicFileInfoProvider fileInfoProvider, DynamicFileInfo fileInfo, ImmutableArray<string> folders)
+            {
+                var documentInfo = CreateDocumentInfoFromFileInfo(fileInfo, folders.IsDefault ? null : (IEnumerable<string>)folders);
+                var documentId = documentInfo.Id;
+
+                lock (_project._gate)
+                {
+                    var filePath = documentInfo.FilePath;
+                    if (_documentPathsToDocumentIds.ContainsKey(filePath))
+                    {
+                        throw new ArgumentException($"'{filePath}' has already been added to this project.", nameof(filePath));
+                    }
+
+                    _documentPathsToDocumentIds.Add(filePath, documentId);
+
+                    _documentIdToDynamicFileInfoProvider.Add(documentId, fileInfoProvider);
+
+                    if (_project._eventSubscriptionTracker.Add(fileInfoProvider))
+                    {
+                        // subscribe to the event when we use this provider the first time
+                        fileInfoProvider.Updated += _project.OnDynamicFileInfoUpdated;
+                    }
+
+                    if (_project._activeBatchScopes > 0)
+                    {
+                        _documentsAddedInBatch.Add(documentInfo);
+                    }
+                    else
+                    {
+                        // right now, assumption is dynamically generated file can never be opened in editor
+                        _project._workspace.ApplyChangeToWorkspace(w => _documentAddAction(w, documentInfo));
+                    }
+                }
+            }
+
+            public IDynamicFileInfoProvider RemoveDynamicFile(string fullPath)
+            {
+                if (string.IsNullOrEmpty(fullPath))
+                {
+                    throw new ArgumentException($"{nameof(fullPath)} isn't a valid path.", nameof(fullPath));
+                }
+
+                lock (_project._gate)
+                {
+                    if (!_documentPathsToDocumentIds.TryGetValue(fullPath, out var documentId) ||
+                        !_documentIdToDynamicFileInfoProvider.TryGetValue(documentId, out var fileInfoProvider))
+                    {
+                        throw new ArgumentException($"'{fullPath}' is not a dynamic file of this project.");
+                    }
+
+                    _documentIdToDynamicFileInfoProvider.Remove(documentId);
+
+                    RemoveFileInternal(documentId, fullPath);
+
+                    return fileInfoProvider;
+                }
+            }
+
             public void RemoveFile(string fullPath)
```

Comment:
this scares the bejesus out of me.  it's a public method, so anyone can call it.  but it's being called by new code that is under a lock.  I don't really understand any of the threading/locking model here.  It *especially* doesn't help that there aren't threading asserts to give me confidence that this is doing the right thing.

Reply:
sure, go and open an issue. if people feels same, we will fix that up.

Reply:
Why not just fix it as part of the PR?  This is a very scary and dangerous approach being taken here.  I don't see why we need an issue to track fixing that.  instead, it would be better to just not introduce the new code *in the first place*.

Or, in other words: this is a PR.  THe point is to ensure that the code is of a sufficient quality level to be merged into the product.  Right now this code is not at that quality level.  Threading/locking are extremely haphazard, and you haven't even been able to describe in general terms why i should even feel confident this is being done properly.  We don't need an issue to track addressing that.  Instead, that should be the *minimum bar* we're expect a PR to meet in order ot be merged into the product.